hoomd/md/manifold.py

```diff
@@ -0,0 +1,389 @@
+# Copyright (c) 2009-2019 The Regents of the University of Michigan
+# This file is part of the HOOMD-blue project, released under the BSD 3-Clause License.
+
+# Maintainer: joaander / All Developers are free to add commands for new features
+
+import hoomd;
+from hoomd.md import _md
+from hoomd import _hoomd
+from hoomd.operation import _HOOMDBaseObject
+from hoomd.data.parameterdicts import ParameterDict, TypeParameterDict
+from hoomd.data.typeconverter import OnlyIf, to_type_converter
+from collections.abc import Sequence
+
+## \internal
+# \brief Base class for manifold
+#
+# A manifold in hoomd reflects a Manifold in c++. It is responsible to define the manifold 
+# used for RATTLE integrators and the active force constraints.
+class Manifold(_HOOMDBaseObject):
+    r"""Base class manifold object.
+
+    Manifold defines a positional constraint to a given set of particles. A manifold can 
+    be applied to a RATTLE method and/or the active force class. The degrees of freedom removed from 
+    the system by constraints are correctly taken into account, i.e. when computing temperature 
+    for thermostatting and/or logging.
+
+    Sphere, Ellipsoid, Plane, Cylinder, Gyroid, Diamond and Primitive are subclasses of this class. All
+    manifolds are described by implicit functions.
+
+    Attention:
+        Users should not instantiate :py:class:`Manifold` directly.
+
+    Warning:
+        Only one manifold can be applied to the methods/active forces.
+
+    """
+    def __init__(self):
+
+        self._cpp_obj = None
+
+        self.name = None
+
+
+class Cylinder(Manifold):
+    R""" Cylinder manifold.
+
+    Args:
+        r (`float`): radius of the cylinder constraint (in distance units).
+        P (`tuple`[ `float`, `float`, `float`]): point defining position of the cylinder axis (default origin).
+
+    :py:class:`Cylinder` specifies that a cylindric manifold is defined as
+    a constraint.
+
+    Note:
+        The cylinder axis is parallel to the z-direction.
+
+    .. rubric:: Implicit function
+
+    .. math::
+        F(x,y,z) = x^{2} + y^{2} - r^{2}
+
+    Example::
+
+        cylinder1 = manifold.Cylinder(r=10)
+        cylinder2 = manifold.Cylinder(r=5,P=(1,1,1))
+    """
+    def __init__(self,r, P=[0,0,0] ):
+        # initialize the base class
+        super().__init__();
+        param_dict = ParameterDict(
+            r=float(r),
+            P=(float, float,float),
+        )
+        param_dict.update(
+            dict(P=(P[0], P[1], P[2])))
+
+        self._param_dict.update(param_dict)
+        # set defaults
+
+    def _attach(self):
+        self._cpp_obj = _md.ManifoldCylinder(self.r, _hoomd.make_scalar3( self.P[0], self.P[1], self.P[2]) );
+
+        self.name = "Cylinder"
+
+        super()._attach()
```

Comment:
`super()._attach()` is attempting to call `setCommunicator` on `self._cpp_obj`. I do not want to make that call conditional because one common type of bug is forgetting to add `setCommunicator` to an object. However, in your case the object has no need of a `Communicator`. One solution I can think of is for you to add an `_attach` method in `Manifold` that does:
```
        self._apply_param_dict()
        self._apply_typeparam_dict(self._cpp_obj, self._simulation)
```
and does *not* call `super().attach()` further up the chain.

The other solution is to not inherit from `_HOOMDBaseObject` and manually add properties for the parameters. Speaking of which, I don't recall seeing the `def_property` calls for the C++ classes to make the parameters accessible. Should the user be able to query or set manifold parameters after construction? If not, then there is no need for the `param_dict` logic from `_HOOMDBaseObject`.